changelog.d/15782.misc

```diff
@@ -0,0 +1 @@
+Add read/write style cross-worker locks.
```

Comment:
I think this looks good :+1: 

I haven't fully mind-melded into the code around the constraints to see something wrong but the tests look robust enough to catch any errant behavior. If you want me to fully understand this, feel free to poke again specifically.

Reply:
Gave the constraints a closer look. I think they make sense.

Kept trying to think if we could do it with one table but couldn't figure anything out that gets us that same read xor write behavior.